tilavarauspalvelu/integrations/keyless_entry/client.py

```diff
@@ -41,124 +51,374 @@ class PindoraClient(BaseExternalServiceClient):
+            action="deleting reservation",
+            expected_status_code=HTTP_204_NO_CONTENT,
+        )
 
     @classmethod
     def change_reservation_access_code(cls, reservation: Reservation) -> PindoraReservationResponse:
-        """Change reservation access code in Pindora."""
+        """Change a reservation's access code in Pindora."""
         url = cls._build_url(f"change-access-code/reservation/{reservation.ext_uuid}")
 
         response = cls.put(url=url)
-        cls._validate_response(response)
+        cls._validate_reservation_response(
+            response,
+            reservation,
+            action="changing access code for reservation",
+        )
 
-        if response.status_code == HTTP_404_NOT_FOUND:
-            msg = f"Reservation '{reservation.ext_uuid}' not found from Pindora."
+        data = cls.response_json(response)
+        return cls._parse_reservation_response(data)
+
+    ####################
+    # Seasonal booking #
+    ####################
+
+    @classmethod
+    def get_seasonal_booking(cls, application_section: ApplicationSection) -> PindoraSeasonalBookingResponse:
+        """Fetch a seasonal booking from Pindora."""
+        url = cls._build_url(f"seasonal-booking/{application_section.ext_uuid}")
+
+        response = cls.get(url=url)
+        cls._validate_seasonal_booking_response(
+            response,
+            application_section,
+            action="fetching seasonal booking",
+        )
+
+        data = cls.response_json(response)
+        return cls._parse_seasonal_booking_response(data)
+
+    @classmethod
+    def create_seasonal_booking(
+        cls,
+        application_section: ApplicationSection,
+        *,
+        is_active: bool = False,
+    ) -> PindoraSeasonalBookingResponse:
+        """Create a new seasonal booking in Pindora."""
+        url = cls._build_url("seasonal-bookings")
+
+        reservations: list[Reservation] = list(
+            application_section.actions.get_reservations().select_related("recurring_reservation__reservation_unit")
+        )
+
+        if not reservations:
+            msg = f"No reservations in for seasonal booking '{application_section.ext_uuid}'."
             raise PindoraAPIError(msg)
 
-        if response.status_code != HTTP_200_OK:
-            msg = (
-                f"Unexpected response from Pindora when changing access code for reservation '{reservation.ext_uuid}': "
-                f"[{response.status_code}] {response.text}"
-            )
+        data = PindoraSeasonalBookingCreateData(
+            seasonal_booking_id=str(application_section.ext_uuid),
+            series=[
+                PindoraSeasonalBookingReservationData(
+                    reservation_unit_id=str(reservation.recurring_reservation.reservation_unit.uuid),
+                    begin=reservation.begin.astimezone(DEFAULT_TIMEZONE).isoformat(),
+                    end=reservation.end.astimezone(DEFAULT_TIMEZONE).isoformat(),
```

Comment:
Could use `date_utils.local_iso_format` for this? _(and for other lines in this same file also)_

Reply:
Might be better, since it leaves out fractional seconds too (not that we have them).